From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Mon, 29 Jun 2026 01:24:13 +0800
Subject: [PATCH 4/5] iommu: apple-dart: Make TCR access hardware-specific

Make TCR access harcware-specific to allow support for DARTs requiring
different logic for TCR access to be added.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 drivers/iommu/apple-dart.c | 88 +++++++++++++++++++++++---------------
 1 file changed, 54 insertions(+), 34 deletions(-)

diff --git a/drivers/iommu/apple-dart.c b/drivers/iommu/apple-dart.c
index ebf4547d3..ff0a08042 100644
--- a/drivers/iommu/apple-dart.c
+++ b/drivers/iommu/apple-dart.c
@@ -160,36 +160,6 @@ enum dart_type {
 	DART_T8110,
 };
 
-struct apple_dart_hw {
-	enum dart_type type;
-	irqreturn_t (*irq_handler)(int irq, void *dev);
-	int (*invalidate_tlb)(struct apple_dart_stream_map *stream_map);
-
-	u32 oas;
-	enum io_pgtable_fmt fmt;
-
-	int max_sid_count;
-
-	u32 lock;
-	u32 lock_bit;
-
-	u32 error;
-
-	u32 enable_streams;
-
-	u32 tcr;
-	u32 tcr_enabled;
-	u32 tcr_disabled;
-	u32 tcr_bypass;
-	u32 tcr_4level;
-
-	u32 ttbr;
-	u32 ttbr_valid;
-	u32 ttbr_addr_field_shift;
-	u32 ttbr_shift;
-	int ttbr_count;
-};
-
 /*
  * Private structure associated with each DART device.
  *
@@ -231,6 +201,38 @@ struct apple_dart {
 	u32 save_ttbr[DART_MAX_STREAMS][DART_MAX_TTBR];
 };
 
+struct apple_dart_hw {
+	enum dart_type type;
+	irqreturn_t (*irq_handler)(int irq, void *dev);
+	int (*invalidate_tlb)(struct apple_dart_stream_map *stream_map);
+	u32 (*read_tcr)(struct apple_dart *dart, u32 sid);
+	void (*write_tcr)(struct apple_dart *dart, u32 sid, u32 val);
+
+	u32 oas;
+	enum io_pgtable_fmt fmt;
+
+	int max_sid_count;
+
+	u32 lock;
+	u32 lock_bit;
+
+	u32 error;
+
+	u32 enable_streams;
+
+	u32 tcr;
+	u32 tcr_enabled;
+	u32 tcr_disabled;
+	u32 tcr_bypass;
+	u32 tcr_4level;
+
+	u32 ttbr;
+	u32 ttbr_valid;
+	u32 ttbr_addr_field_shift;
+	u32 ttbr_shift;
+	int ttbr_count;
+};
+
 /*
  * Convenience struct to identify streams.
  *
@@ -331,7 +333,17 @@ apple_dart_hw_enable_translation(struct apple_dart_stream_map *stream_map, int l
 	WARN_ON(levels != 3 && levels != 4);
 	WARN_ON(levels == 4 && !dart->four_level);
 	for_each_set_bit(sid, stream_map->sidmap, dart->num_streams)
-		writel(tcr, dart->regs + DART_TCR(dart, sid));
+		dart->hw->write_tcr(dart, sid, tcr);
+}
+
+static u32 apple_dart_t8020_read_tcr(struct apple_dart *dart, u32 sid)
+{
+	return readl(dart->regs + DART_TCR(dart, sid));
+}
+
+static void apple_dart_t8020_write_tcr(struct apple_dart *dart, u32 sid, u32 val)
+{
+	writel(val, dart->regs + DART_TCR(dart, sid));
 }
 
 static void apple_dart_hw_disable_dma(struct apple_dart_stream_map *stream_map)
@@ -340,7 +352,7 @@ static void apple_dart_hw_disable_dma(struct apple_dart_stream_map *stream_map)
 	int sid;
 
 	for_each_set_bit(sid, stream_map->sidmap, dart->num_streams)
-		writel(dart->hw->tcr_disabled, dart->regs + DART_TCR(dart, sid));
+		dart->hw->write_tcr(dart, sid, dart->hw->tcr_disabled);
 }
 
 static void
@@ -1280,6 +1292,8 @@ static const struct apple_dart_hw apple_dart_hw_t8103 = {
 	.type = DART_T8020,
 	.irq_handler = apple_dart_t8020_irq,
 	.invalidate_tlb = apple_dart_t8020_hw_invalidate_tlb,
+	.read_tcr = apple_dart_t8020_read_tcr,
+	.write_tcr = apple_dart_t8020_write_tcr,
 	.oas = 36,
 	.fmt = APPLE_DART,
 	.max_sid_count = 16,
@@ -1306,6 +1320,8 @@ static const struct apple_dart_hw apple_dart_hw_t8103_usb4 = {
 	.type = DART_T8020,
 	.irq_handler = apple_dart_t8020_irq,
 	.invalidate_tlb = apple_dart_t8020_hw_invalidate_tlb,
+	.read_tcr = apple_dart_t8020_read_tcr,
+	.write_tcr = apple_dart_t8020_write_tcr,
 	.oas = 36,
 	.fmt = APPLE_DART,
 	.max_sid_count = 64,
@@ -1332,6 +1348,8 @@ static const struct apple_dart_hw apple_dart_hw_t6000 = {
 	.type = DART_T6000,
 	.irq_handler = apple_dart_t8020_irq,
 	.invalidate_tlb = apple_dart_t8020_hw_invalidate_tlb,
+	.read_tcr = apple_dart_t8020_read_tcr,
+	.write_tcr = apple_dart_t8020_write_tcr,
 	.oas = 42,
 	.fmt = APPLE_DART2,
 	.max_sid_count = 16,
@@ -1358,6 +1376,8 @@ static const struct apple_dart_hw apple_dart_hw_t8110 = {
 	.type = DART_T8110,
 	.irq_handler = apple_dart_t8110_irq,
 	.invalidate_tlb = apple_dart_t8110_hw_invalidate_tlb,
+	.read_tcr = apple_dart_t8020_read_tcr,
+	.write_tcr = apple_dart_t8020_write_tcr,
 	.fmt = APPLE_DART2,
 	.max_sid_count = 256,
 
@@ -1386,7 +1406,7 @@ static __maybe_unused int apple_dart_suspend(struct device *dev)
 	unsigned int sid, idx;
 
 	for (sid = 0; sid < dart->num_streams; sid++) {
-		dart->save_tcr[sid] = readl(dart->regs + DART_TCR(dart, sid));
+		dart->save_tcr[sid] = dart->hw->read_tcr(dart, sid);
 		for (idx = 0; idx < dart->hw->ttbr_count; idx++)
 			dart->save_ttbr[sid][idx] =
 				readl(dart->regs + DART_TTBR(dart, sid, idx));
@@ -1411,7 +1431,7 @@ static __maybe_unused int apple_dart_resume(struct device *dev)
 		for (idx = 0; idx < dart->hw->ttbr_count; idx++)
 			writel(dart->save_ttbr[sid][idx],
 			       dart->regs + DART_TTBR(dart, sid, idx));
-		writel(dart->save_tcr[sid], dart->regs + DART_TCR(dart, sid));
+		dart->hw->write_tcr(dart, sid, dart->save_tcr[sid]);
 	}
 
 	return 0;
